src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs

```diff
@@ -893,53 +901,43 @@ private bool ConsumeStringNextSegment()
                                                 byte nextByte = localBuffer[j];
                                                 if (!JsonReaderHelper.IsHexDigit(nextByte))
                                                 {
-                                                    _currentPosition = startPosition;
+                                                    RollBackStateWhenParsingString(rollBackState, isError: true);
                                                     ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.InvalidHexCharacterWithinString, nextByte);
                                                 }
-                                                if (j - idx > numberOfHexDigits)
+                                                numberOfHexDigits++;
+                                                _bytePositionInLine++;
+                                                if (numberOfHexDigits >= 4)
                                                 {
-                                                    if (movedToNext)
-                                                    {
-                                                        nextCharEscaped = false;
-                                                        goto StartOfLoop;
-                                                    }
-                                                    goto DoneWithHex;
+                                                    nextCharEscaped = false;
+                                                    idx = j + 1; // Skip the 4 hex digits, the for loop accounts for idx incrementing past the 'u'
+                                                    goto StartOfLoop;
                                                 }
-                                                _bytePositionInLine++;
                                             }
 
                                             if (!GetNextSpan())
                                             {
-                                                _currentPosition = startPosition;
                                                 if (IsLastSpan)
                                                 {
+                                                    RollBackStateWhenParsingString(rollBackState, isError: true);
                                                     ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.EndOfStringNotFound);
                                                 }
 
                                                 // We found less than 4 hex digits.
-                                                _lineNumber = prevLineNumber;
-                                                _bytePositionInLine = prevPosition;
-                                                _totalConsumed = prevTotalConsumed;
+                                                RollBackStateWhenParsingString(rollBackState);
```

Comment:
Some of the cases have different logic so they can't benefit from this. Given the `IsLastSpan` property can change between calls to `GetNextSpan()`, I am OK with the larger IL for clarity. Only 4 instances of the ~20 call pairs would benefit from this, so I am going to leave it as is and not use IsLastSpan as a surrogate for "error".

I was in fact thinking of having two separate roll back methods rather than forcing a branch on isError in the "happy" path of valid JSON but not enough data.